### PR TITLE
fix issue 20761 - __traits(isSame) for alias tuples is broken and und…

### PR DESCRIPTION
**Pull request by ghost**
…erspecified

### CODE REVIEW

File: src/dmd/traits.d [hunk 1]
```diff
@@ -1694,12 +1694,35 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
             return new ErrorExp();
 
+        // example:  __traits(isSame, seq!(0,0), seq(1,1,1)  -> 2 != 3
+        if (ob1.dim != ob2.dim)
+            return False();
+        // expansion has happened
+        else if (ob1.dim > 1)
+        {
+            foreach (immutable i; 0 .. ob1.dim)
+                if (!ob1[i].equals(ob2[i]))
```
Comment (MoonlightSentinel) [inline src/dmd/traits.d:1719]: Don't think this should use `equals` directly and instead apply `isSame` pairwise to all elements:

Example:
```
static assert(__traits(isSame, i => i.value, a => a.value)); // Passes

static assert(__traits(isSame, (a, b) => a + b, (x, y) => x + y)); ; // Passes

static assert(__traits(isSame, // Fails
    Seq!(
        i => i.value,
        (a, b) => a + b
    ),
    Seq!(
        a => a.value,
        (x, y) => x + y
    )
));
```

TL:DR:
`__traits(isSame(Seq!(x0, x1, ...), Seq!(y0, y1, ...)))` should be equal  to
`__traits(isSame(x0, y0)) && __traits(isSame(x1, y1)) && ....`

Comment (NilsLankila) [reply src/dmd/traits.d:1719]: yeah, I know. The current implementation needs to be refactrored... something like bool `isSame (RootObject, RootObject, Scope* sc);`, then I can use this func on each individual element, i.e after expansion.

Comment (NilsLankila) [reply src/dmd/traits.d:1719]: the call to `equals` does not cover the specials cases for each individual element

Comment (MoonlightSentinel) [reply src/dmd/traits.d:1719]: Then i would suggest to split this PR into two commits:
1. Refactor the current logic into a nested function
2. Apply your fix using the introduced function

Comment (NilsLankila) [reply src/dmd/traits.d]: no need to split

**Comment by NilsLankila** [comment]
This good now. All previous fix made to isSame are applied to individual tuple elements.

File: src/dmd/traits.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/traits.d]: Refactorings and bug fixes should always be split in different commits s.t. it's easier to track down potential errors later , e.g. when using digger. (Also greatly improves the reviewability of a PR)

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]
Mssing https://github.com/dlang/dmd/pull/11373#discussion_r450449166 but otherswise LGTM

**Comment by NilsLankila** [comment]
okay done

**Comment by NilsLankila** [comment]
wait... the TC has slipped in the first commit

**Review by MoonlightSentinel — DISMISSED** [review]

File: src/dmd/traits.d [hunk 2]
```diff
@@ -1989,3 +1880,128 @@ Lnext:
         e.error("unrecognized trait `%s`", e.ident.toChars());
     return new ErrorExp();
 }
+
+// compare arguments of __traits(isSame)
```
Comment (MoonlightSentinel) [inline src/dmd/traits.d:1884]:
```suggestion
/// compare arguments of __traits(isSame)
```

Comment (NilsLankila) [reply src/dmd/traits.d:1884]: it's private so more a code comment than a documentation one

**Review by MoonlightSentinel — APPROVED** [review]
Thanks

**Comment by NilsLankila** [comment]
alright, I'm no longer answerable for anything now.

File: src/dmd/traits.d [hunk 3]
```diff
@@ -1694,133 +1694,24 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
 
-        if (!TemplateInstance.semanticTiargs(e.loc, sc, e.args, 0))
+        // https://issues.dlang.org/show_bug.cgi?id=20761
+        // tiarg semantic may expand in place the list of arguments, for example:
+        //
+        //     before tiarg sema:  __traits(isSame, seq!(0,0), seq!(1,1))
+        //     after            :  __traits(isSame, 0, 0, 1, 1)
+        //
+        // so we split in two lists
+        Objects ob1; ob1.push((*e.args)[0]);
```
Comment (UplinkCoder) [inline src/dmd/traits.d:1704]: that's not dmd style.

Comment (MoonlightSentinel) [reply src/dmd/traits.d]: Good catch. Fixed

**Comment by NilsLankila** [comment]
label's gone again

**Comment by wilzbach** [comment]
Why did no one ask for a test to be added? Tests should be the bare minimum!!

**Comment by MoonlightSentinel** [comment]
Because the original version had tests which apparenty got lost during the commit partitionin (and I'm blind as well).

Added some tests in #11392

**Comment by NilsLankila** [comment]
test got lost here 

> NilsLankila force-pushed the NilsLankila:issue-20761 branch from 6f79913 to a637577 7 hours ago

**Pull request merged**